src/Components/Server/src/Circuits/RemoteJSDataStream.cs

```diff
@@ -0,0 +1,165 @@
+            RemoteJSRuntime runtime,
+            IJSDataReference jsDataReference,
+            long totalLength,
+            long maxBufferSize,
+            long maximumIncomingBytes,
+            CancellationToken cancellationToken = default)
+        {
+            var streamId = Guid.NewGuid();
+            var remoteJSDataStream = new RemoteJSDataStream(runtime, streamId, totalLength, maxBufferSize, cancellationToken);
+            await runtime.InvokeVoidAsync("Blazor._internal.sendJSDataStream", jsDataReference, streamId, maximumIncomingBytes);
+            return remoteJSDataStream;
+        }
+
+        private RemoteJSDataStream(
+            RemoteJSRuntime runtime,
+            Guid streamId,
+            long totalLength,
+            long maxBufferSize,
+            CancellationToken cancellationToken)
+        {
+            _runtime = runtime;
+            _streamId = streamId;
+            _totalLength = totalLength;
+            _cancellationToken = cancellationToken;
+
+            _runtime.RemoteJSDataStreamInstances.Add(_streamId, this);
+
+            _pipe = new Pipe(new PipeOptions(pauseWriterThreshold: maxBufferSize, resumeWriterThreshold: maxBufferSize));
+            _pipeReaderStream = _pipe.Reader.AsStream();
+        }
+
+        // Ideally we'd have IAsyncEnumerable<ReadOnlySequence<byte>> here so we can pass through the
+        // data without having to copy it into a temporary buffer in BlazorPackHubProtocolWorker. But trying
+        // this gives strange errors; sometimes the "chunk" variable below has a negative length, even though
+        // the logic never returns a corrupted item as far as I can tell.
+        private async Task ReceiveData(byte[] chunk, string error)
+        {
+            try
+            {
+                if (!string.IsNullOrEmpty(error))
+                {
+                    throw new InvalidOperationException($"An error occurred while reading the remote stream: {error}");
+                }
+
+                if (chunk.Length == 0)
+                {
+                    throw new InvalidOperationException($"The incoming data chunk cannot be empty.");
+                }
+
+                _bytesRead += chunk.Length;
+
+                if (_bytesRead > _totalLength)
+                {
+                    throw new InvalidOperationException($"The incoming data stream declared a length {_totalLength}, but {_bytesRead} bytes were read.");
+                }
+
+                CopyToPipeWriter(chunk, _pipe.Writer);
+                _pipe.Writer.Advance((int)chunk.Length);
+                await _pipe.Writer.FlushAsync(_cancellationToken);
+
+                if (_bytesRead == _totalLength)
+                {
+                    await _pipe.Writer.CompleteAsync();
+                }
+            }
+            catch (Exception e)
+            {
+                await _pipe.Writer.CompleteAsync(e);
+                return;
```

Comment:
```suggestion
```

Reply:
N/A anymore given this will need to return `false`. Previously we were failing silently and the JS side would just keep supplying (unusable) chunks. Given that we're closing the pipe, I believe we should also be disposing of the stream here.